rig/routing_table/minimise.py

```diff
@@ -0,0 +1,112 @@
+import collections
+from rig.routing_table import MinimisationFailedError
+from rig.routing_table.remove_default_routes import minimise as \
+    remove_default_entries
+from rig.routing_table.ordered_covering import minimise as ordered_covering
+from six import iteritems
+
+
+def minimise(routing_tables, target_lengths,
+             methods=(remove_default_entries, ordered_covering)):
+    """
+
+    Parameters
+    ----------
+    routing_tables : {(x, y): [RoutingTableEntry, ...], ...}
+        Dictionary mapping co-ordinates to the routing tables associated with
+        those co-ordinates. NOTE: This is the data structure as returned by
+        :py:meth:`~rig.place_and_route.build_full_routing_tables`.
+    target_lengths : int or {(x, y): int or None, ...} or None
+        Maximum length of routing tables. If an integer this is assumed to be
+        the maximum length for any table; if a dictionary then it is assumed to
+        be a mapping from co-ordinate to maximum length (or None); if None then
+        tables will be minimised as far as possible.
+    methods :
+        Iterable of minimisation methods.
```

Comment:
Each method is tried in the order presented and the first to meet the required target length for a given chip is used. Consequently less computationally costly algorithms should be nearer the start of the list. The defaults will try to remove default routes (:py:meth:`rig.routing_table.remove_default_routes.minimise`) and then fall back on the ordered covering algorithm (:py:meth:`rig.routing_table.ordered_covering.minimise`).